### PR TITLE
Add some runtime checking for unions

### LINKED ISSUE #571: Raise error when using Scalars (or non strawberry types) with Union
If we do something any of the following:

```python
X = strawberry.Union("X", (int, str))

@strawberry.type
class A:
    field: Union[int, str]


@strawberry.type
class A:
    field: Union[NonStrawberryTypeA, NonStrawberryTypeB]
```

we should get an error informing that:

1. you can't use unions with scalars (not supported by GraphQL)
2. you can only use union with strawberry types

### PR DESCRIPTION
**Pull request by jkimbo**
## Description

Fixes #571

Not sure whether it's worth creating specific Strawberry exceptions for this cases. What do you think @patrick91 ?

## Types of Changes

- [x] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [ ] Documentation

## Issues Fixed or Closed by This PR

* #571

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
- [x] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by botberry** [comment]
Hi, thanks for contributing to Strawberry 🍓!

We noticed that this PR is missing a `RELEASE.md` file. We use that to automatically do releases here on GitHub and, most importantly, to PyPI!

So as soon as this PR is merged, a release will be made 🚀.

Here's an example of `RELEASE.md`:

```markdown
Release type: patch

Description of the changes, ideally with some examples, if adding a new feature.
```

Release type can be one of patch, minor or major. We use [semver](https://semver.org/),so make sure to pick the appropriate type. If in doubt feel free to ask :)

**Comment by jkimbo** [comment]
Not sure why codecov thinks this change will reduce code coverage...

File: tests/types/test_union.py [hunk 1]
```diff
@@ -138,3 +139,24 @@ class B:
 
     with pytest.raises(ValueError, match=r"Cannot use union type directly"):
         Result()
+
+
+def test_error_with_empty_type_list():
+    with pytest.raises(Exception, match="No types passed to `union`"):
```
Comment (BryceBeagle) [inline tests/types/test_union.py:145]: TIL `match` argument. I will have to use this in the future.

**Review by BryceBeagle — CHANGES_REQUESTED** [review]
Should we create custom exceptions (i.e. in `strawberry.exceptions`) for these?

**Comment by jkimbo** [comment]
> Should we create custom exceptions (i.e. in `strawberry.exceptions`) for these?

I was wondering about that. Happy to do so though I can't see developers wanting to catch specific exceptions. Also I think the `len(types)` check should probably stay as a TypeError.

**Comment by patrick91** [comment]
> Should we create custom exceptions (i.e. in `strawberry.exceptions`) for these?

Good question!

My idea around exception was that at some point we'd have a way to link to the docs for some of them[1], especially the ones that maybe don't make a lot of sense to a beginner :)

[1] Via a custom __repr__ or something :)

**Comment by BryceBeagle** [comment]
> Also I think the len(types) check should probably stay as a TypeError

Technically speaking, our custom exception could just extend `TypeError`

**Comment by jkimbo** [comment]
> My idea around exception was that at some point we'd have a way to link to the docs for some of them[1], especially the ones that maybe don't make a lot of sense to a beginner :)

I like this reasoning around when to add custom exceptions. Will update the PR with some.

**Review by patrick91 — APPROVED** [review]
Perfect, I was struggling understanding why we changed the test, but the I noticed the test was wrong 🤣 nice one!

**Pull request merged**